src/run-cache.test.ts

```diff
@@ -236,93 +272,280 @@ describe("RunCache", () => {
-          expect(cacheState.value).toBe(JSON.stringify("updatedValue"));
+          expect(cacheState.key).toStrictEqual(key);
+          expect(cacheState.value).toStrictEqual(JSON.stringify(value));
+          expect(cacheState.ttl).toStrictEqual(100);
         },
       );
 
+      RunCache.onExpiry(funcToBeExecutedOnExpiry);
+      RunCache.onKeyExpiry(key, funcToBeExecutedOnExpiry);
+
+      RunCache.set({
+        key,
+        value: value,
+        ttl: 100,
+      });
+
+      jest.advanceTimersByTime(101);
+      await Promise.resolve(); // Flush microtasks
+
+      expect(funcToBeExecutedOnExpiry).toHaveBeenCalledTimes(2);
+    });
+  });
+
+  describe("onRefetch() and onKeyRefetch()", () => {
+    it("should trigger on refetch", async () => {
+      const key = uuid();
+      let dynamicValue = uuid();
+
+      const funcToBeExecutedOnRefetch = jest.fn((cacheState: EventParam) => {
+        expect(cacheState.key).toStrictEqual(key);
+        expect(cacheState.value).toStrictEqual(JSON.stringify(dynamicValue));
+        expect(cacheState.ttl).toStrictEqual(100);
+      });
+
+      const sourceFn = jest.fn(() => dynamicValue);
+
+      RunCache.onRefetch(funcToBeExecutedOnRefetch);
+      RunCache.onKeyRefetch(key, funcToBeExecutedOnRefetch);
+
       await RunCache.set({
-        key: "key2",
+        key,
         sourceFn,
-        onRefetch: funcToBeExecutedOnRefetch,
+        ttl: 100,
+        autoRefetch: true,
       });
 
+      dynamicValue = uuid();
+
+      jest.advanceTimersByTime(101);
+
+      await Promise.resolve(); // Flush microtasks
+
+      expect(sourceFn).toHaveBeenCalledTimes(2);
+      expect(funcToBeExecutedOnRefetch).toHaveBeenCalledTimes(2);
+    });
+  });
+
+  describe("onRefetchFailure() and onKeyRefetchFailure()", () => {
+    it("should trigger if the sourceFn fails", async () => {
+      const key = uuid();
+      const value = uuid();
+
+      let breaker = false;
+
+      const funcToBeExecutedOnRefetchFailure = jest.fn(
+        (cacheState: EventParam) => {
+          expect(cacheState.key).toStrictEqual(key);
+          expect(cacheState.value).toStrictEqual(JSON.stringify(value));
+          expect(cacheState.ttl).toStrictEqual(100);
+        },
+      );
+
+      const sourceFn = jest.fn(() => {
+        if (breaker) {
+          throw Error("Simulated source function failure");
+        } else {
+          return uuid();
+        }
+      });
+
+      RunCache.onRefetchFailure(funcToBeExecutedOnRefetchFailure);
+      RunCache.onKeyRefetchFailure(key, funcToBeExecutedOnRefetchFailure);
+
+      await RunCache.set({
+        key,
+        value,
+        sourceFn,
+        ttl: 100,
+        autoRefetch: true,
+      });
+
+      breaker = true;
+
+      jest.advanceTimersByTime(101);
+
+      await Promise.resolve(); // Flush microtasks
+
       expect(sourceFn).toHaveBeenCalledTimes(1);
+      expect(funcToBeExecutedOnRefetchFailure).toHaveBeenCalledTimes(2);
+    });
```

Comment:
The `sourceFn` is not being called at `set` because a value is provided. `funcToBeExecutedOnRefetchFailure` is called twice because we have it set on two event triggers.

```ts
RunCache.onRefetchFailure(funcToBeExecutedOnRefetchFailure);
RunCache.onKeyRefetchFailure(key, funcToBeExecutedOnRefetchFailure);
```